### PR TITLE
refactor(accesscontrol): deterministic cache key hashing

### PR DESCRIPTION
**Pull request by aruiz14**
### Problem
The current `AccessFor` implementation calculates the cache key for a given user based on the resource version of Roles/ClusterRoles associated via RoleBindings/ClusterRoleBindings. While this is correct, in the case of a cache miss that information that was previously used to calculate the hash is no used but instead retrieved again. Finally, the previously-calculated hash is used as key to store the generated AccessSet into the cache.

This approach Introduces the chance for a race condition, which this PR is meant to fix by using the same data used to calculate the hash to later generate the AccessSet in case of a cache miss.

### Summary
* Small refactoring of some internal functions to allow reusing them
* Add new types for encapsulating all needed data
* Make accessStore use the new types and method
* Cleanup unused code and adapt tests to new signatures
* Add new comments to structs and methods
* Refactor PolicyRules' indexer to make it easier to read and more testable

(please take a look at the [commits view](https://github.com/rancher/steve/pull/292/commits) for more details)

### CODE REVIEW

File: pkg/accesscontrol/policy_rule_index.go [hunk 1]
```diff
@@ -108,23 +109,24 @@ func (p *policyRuleIndex) addAccess(accessSet *AccessSet, namespace string, role
 	}
 }
 
-func (p *policyRuleIndex) getRules(namespace string, roleRef rbacv1.RoleRef) []rbacv1.PolicyRule {
+// getRules obtain the actual Role or ClusterRole pointed at by a RoleRef, and returns PolicyRyles and the resource version
```
Comment (ericpromislow) [inline pkg/accesscontrol/policy_rule_index.go:112]: Typo: should be `PolicyRules`

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Looks fine, apart from a trivial typo

**Review by ericpromislow — DISMISSED** [review]
thx

**Review by bigkevmcd — DISMISSED** [review]
I've gone over this in fine detail, and it looks fine.

It improves the test coverage a fair bit, but there's still a lot of untested code that we can address when we next touch these files.

File: pkg/accesscontrol/access_store.go [hunk 2]
```diff
@@ -42,21 +33,14 @@ type accessStoreCache interface {
...
 }
 
-type roleKey struct {
-	namespace string
-	name      string
-}
-
-func NewAccessStore(ctx context.Context, cacheResults bool, rbac v1.Interface) *AccessStore {
+func NewAccessStore(_ context.Context, cacheResults bool, rbac v1.Interface) *AccessStore {
 	as := &AccessStore{
```
[the remark this answers is not in the document]
Comment (aruiz14) [reply pkg/accesscontrol/access_store.go:41]: I did not want to break the public interface

**Comment by aruiz14** [comment]
@ericpromislow please take another look, I've addressed Kevin's feedback but it didn't affect core of this PR.

**Review by bigkevmcd — APPROVED** [review]

**Review by ericpromislow — APPROVED** [review]
Works and looks ok

**Pull request merged**